### PR TITLE
`Development`: Refactor DatabaseUtilService and ModelFactory

### PR DESCRIPTION
**Pull request by DominikRemo**
#### General

- [x] The changed server tests succeed locally and on bamboo.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context
The DatabaseUtilService and ModelFactory have become complex over time. To address this, this and two other Pull Requests focus on splitting both classes into smaller, more specific Factories and UtilServices. The goal is to improve maintainability and extensibility.

By breaking down both classes, we aim to simplify their functionality and make them easier to manage. The modular approach enhances code maintainability and readability. It also facilitates future extensions by providing more explicit boundaries and promoting reusability.

In summary, the PRs aim to split the DatabaseUtilService and ModelFactory into more specific components, enabling easier maintenance and extension. This refactoring will result in a more streamlined codebase, benefiting overall development efforts.

Note that our three PRs only split the methods into new and more specific classes. The aim of this PR is not to change or improve these methods.

### Description
_This is one of three PRs related to refactoring the DatabaseUtilService and ModelFactory._
In this PR, new UtilServices and Factories have been created. The methods in DatabaseUtilService and ModelFactory were commented and copied to the new classes accordingly.

### Steps for Testing
**Test changes only**

#### Code Review
- [x] Code Review 1

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseUtilService.java [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+    /**
+     * Add a submission with a result to the given programming exercise. The submission will be assigned to the corresponding participation of the given login (if exists or
+     * create a new participation).
+     * The method will make sure that all necessary entities are connected.
+     *
+     * @param exercise   for which to create the submission/participation/result combination.
+     * @param submission to use for adding to the exercise/participation/result.
+     * @param login      of the user to identify the corresponding student participation.
+     */
+    public void addProgrammingSubmissionWithResult(ProgrammingExercise exercise, ProgrammingSubmission submission, String login) {
```
Comment (laadvo) [inline src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseUtilService.java:433]: I am not sure if this method should be here (ProgrammingExerciseUtilService) or in the ParticipationUtilService. There are some similar methods in the ParticipationUtilService

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java [hunk 2]
```diff
@@ -3312,6 +3882,9 @@ public ModelingSubmission addModelingSubmissionWithEmptyResult(ModelingExercise
         return submission;
     }
 
+    /**
+     * Moved to {@link ModelingExerciseUtilService#addModelingSubmission(ModelingExercise, ModelingSubmission, String)}
+     */
     public ModelingSubmission addModelingSubmission(ModelingExercise exercise, ModelingSubmission submission, String login) {
```
Comment (laadvo) [inline src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java:3888]: here as well, I am not sure if it maybe belongs in ParticipationUtilService instead

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java [hunk 3]
```diff
@@ -3871,6 +4589,7 @@ public ExampleSubmission generateExampleSubmission(String modelOrText, Exercise
     }
 
     /**
+     * Moved to {@link QuizExerciseUtilService#generateSubmittedAnswerFor(QuizQuestion, boolean)}
```
Comment (laadvo) [inline src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java:4592]: this and many other methods below related to creating quizes should be moved into the QuizFactory instead

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java [hunk 4]
```diff
@@ -3270,6 +3836,9 @@ public Result generateResult(Submission submission, User assessor) {
         return result;
     }
 
+    /**
+     * Moved to {@link ParticipationUtilService#generateComplaintAndResponses(String, int, int, int, boolean, Result, User)}
```
Comment (laadvo) [inline src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java:3840]: I'm not sure if methods related complaints belong in the ParticipationUtilService. Assessment is part of the ExerciseUtilService so it might make sense to also have complaints there. Or what do you think?

**Review by laadvo — CHANGES_REQUESTED** [review]
The refactoring already looks really good! Here are a few suggestions that I had and you might want to consider

File: src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseUtilService.java (hunk 1 above)
Comment (DominikRemo) [reply src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseUtilService.java:434]: Discussed:
Methods related to submissions and specific Exercises go into the corresponding ExerciseUtilServices.
Submissions with no specific Exercise belong into ParticipationUtilService

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java (hunk 3 above)
Comment (DominikRemo) [reply src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java]: Discussed: Done in FollowUps

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java (hunk 4 above)
Comment (DominikRemo) [reply src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java]: Discussed: I will move them into a new ComplaintUtilService

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java [hunk 5]
```diff
@@ -2691,10 +3165,17 @@ public Course addCourseWithNamedProgrammingExercise(String programmingExerciseTi
         return courseRepo.findByIdWithExercisesAndLecturesElseThrow(course.getId());
     }
 
+    /**
+     * Moved to {@link ProgrammingExerciseUtilService#populateProgrammingExercise(ProgrammingExercise, String, String, boolean)}
```
[the remark this answers is not in the document]
Comment (DominikRemo) [reply src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java]: I will move this and the following Method to the ProgrammingExerciseFactory
